src/Groups/directproducts.jl

```diff
@@ -0,0 +1,486 @@
+   @assert length(V)==G.n "Vector of wrong length"
+   x = prod([embedding(Gfull,j)(V[j]) for j in 1:length(V)])
+   @assert x in G "Element not in the group"
+   return group_element(G,x.X)
+end
+
+function (G::DirectProductOfGroups)(V::GAPGroupElem...)
+   Gfull=DirectProductOfGroups(G.Xfull,G.L,G.n,G.Xfull,true)
+   @assert length(V)==G.n "Vector of wrong length"
+   x = prod([embedding(Gfull,j)(V[j]) for j in 1:length(V)])
+   @assert x in G "Element not in the group"
+   return group_element(G,x.X)
+end
+
+# start part on subgroups
+function _as_subgroup_bare(G::DirectProductOfGroups, H::GapObj)
+  t = H==G.X
+  return DirectProductOfGroups(H, G.L, G.n, G.X, t)
+end
+
+function _as_subgroup(G::DirectProductOfGroups, H::GapObj, ::Type{U}) where U
+  H1 = _as_subgroup_bare(G, H)
+  return H1, hom(H1, G, x::U -> group_element(G, x.X))
+end
+
+function _as_subgroup(G::DirectProductOfGroups, H::GapObj)
+  return _as_subgroup(G, H, GAPGroupElem{DirectProductOfGroups})
+end
+
+function sub(G::DirectProductOfGroups, elms::Vector{GAPGroupElem{DirectProductOfGroups}})
+  elems_in_GAP = GAP.julia_to_gap(GapObj[x.X for x in elms])
+  H = GAP.Globals.Group(elems_in_GAP)
+  #H is the group. I need to return the inclusion map too
+  return _as_subgroup(G, H)
+end
+
+function sub(L::GAPGroupElem{DirectProductOfGroups}...)
+   if length(L)==0 throw(ArgumentError("Empty list")) end
+   l=collect(L)
+   @assert all(x -> parent(x) == parent(l[1]), l)
+   return sub(parent(l[1]),l)
+end
+
+function sub(L::Vector{GAPGroupElem{DirectProductOfGroups}})
+   if length(L)==0 throw(ArgumentError("Empty list")) end
+   l=collect(L)
+   @assert all(x -> parent(x) == parent(l[1]), l)
+   return sub(parent(l[1]),l)
+end
+
+function Base.show(io::IO, G::DirectProductOfGroups)
+   if G.isfull
+      print(io, "DirectProduct of ")
+      display(G.L)
+   else
+      print(io, GAP.gap_to_julia(GAP.Globals.StringViewObj(G.X)))
+   end
+end
+
+#=
+function Base.show(io::IO, x::GAPGroupElem{DirectProductOfGroups{S,T}}) where { S, T }
+   print(io, "DirectProductElement( ", GAP.gap_to_julia(GAP.Globals.StringViewObj(projection(parent(x),1)(x).X)), " , ", GAP.gap_to_julia(GAP.Globals.StringViewObj(projection(parent(x),2)(x).X)), " )")
+end
+=#
+
+# if a subgroup of a direct product of groups is also a direct product of groups
+"""
+    write_as_full(G::DirectProductOfGroups)
+If `G` is a subgroup of the direct product `G_1 x ... x G_n` such that `G = H_1 x ... x H_n` for `H_i` subgroup of `G_i`, return `G` as full direct product of the `H_i`. If such `H_i` do not exist, an ERROR is returned.
+"""
+function write_as_full(G::DirectProductOfGroups)
+   if G.isfull
+      return G
+   else
+      LK = [image(projection(G,j))[1] for j in 1:G.n]
+      H = direct_product(LK)
+# index(H,G)==1 does not work because it does not recognize G as a subgroup of H
+      if order(H)==order(G)
+         return H
+      else
+         throw(ArgumentError("G is not a direct product of groups"))
```

Comment:
Yeah, that's the general case of course.

What's the usecase for `write_as_full`? Unless you have one, I'd rather drop this function.

Reply:
I personally find this function useful, since it always says whether a subgroup can be written as a full direct product of subgroups of the factors. Once ascertained to be true, this will allow better working for functions like `show`, `projection`, `embedding`, etc.
But maybe I'm working too much with fantasy :-)

Reply:
How does this make `show`, `projection`, etc. "better"? Could you perhaps explain with a mock example?

I am not saying you are wrong, but indeed, to me these are all fantasy examples, as in practice I never had use for them when working with direct products in research code; perhaps they are useful when teaching, though I still don't quite see how. Hence my question.

Anyway, while I am happy to accommodate teaching and "easy for non-computation experts" when we can do so with moderate effort, I really think we must not compromise the possibility to write efficient code. This is why I am against always performing the "fullness" test, even if the user doesn't need it: by doing so, it is basically guaranteed that this code cannot be used in highly tuned research code.

Reply:
For example: take `G1`, `G2` groups, `H1` and `H2` subgroups of `G1`, `G2` respectively. If I create the subgroup `H1xH2` of `G1xG2`, Oscar considers `H1xH2` not full, because it is a subgroup of `G1xG2`. So, if I want to compute the first projection of `H1xH2` , Oscar computes the first projection `p` of `G1xG2` (through GAP), the embedding `e` of `H1xH2` into `G1xG2` and return the function `e*p`. After turning `H1xH2` into `write_as_full(H1xH2)`, Oscar considers now `H1xH2` as a full direct product, and the underlying GAP object is also the direct product `H1xH2`; so, now it computes directly the first projection of `H1xH2`, without passing through `G1xG2`.
Do you think it is worth?